### ISSUE TITLE
[GETUSHA - WAITING ON CHECKLIST] [$250] [New feature] Add loading indicator when ReconnectApp is running

### ISSUE THREAD

**Issue by slafortune** [Expensify/App#46611]
Original post - 

> **Version Number:** v1.4.77-11
> **Reproducible in staging?:** Yes 
> **Reproducible in production?:** Yes
> **If this was caught during regression testing, add the test name, ID and link from TestRail:**
> **Email or phone of affected tester (no customers):** NA
> **Logs:** https://stackoverflow.com/c/expensify/questions/4856
> **Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/350373
> **Issue reported by:**  @flodnv 
> **Slack conversation:** Original proposal here
> : https://expensify.slack.com/archives/C01GTK53T8Q/p1699972596342559

Sometimes, the app is loading something for a little while; as an end user, you have no clue what's happening. Then, when loading finishes, sometimes several seconds later, all of a sudden, new data pops up and takes you by surprise, resulting in a very unpleasant UX that almost feels broken.

Coming from [this](https://github.com/Expensify/App/issues/43453) we've landed on showing the thin bar loader at the top of the screen when `ReconnectApp` is running.

![image](https://github.com/user-attachments/assets/92e67c13-2ce6-4fa5-8e03-05cc17ad9be9)

@dubielzyk-expensify  created a [CodePen](https://codepen.io/dubielzyk/full/ExBVxVG) and it's the animation called Rubber Band or #loading1 in code. 

> My code isn't the best so I'd urge us to optimize it rather than copy paste 😅

Much discussion took place [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1699972596342559)

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021836556152855644897</li>
        <li>Upwork Job ID: 1836556152855644897</li>
        <li>Last Price Increase: 2024-09-19</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Contributor | 104442782</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @jliexpensify</details>

**Comment by getusha** [comment]
@slafortune could you assign me here too? thanks :)

**Comment by jliexpensify** [comment]
Assigned you @getusha

**Comment by dubielzyk-expensify** [comment]
Excited about this one. Let me know when there's something to look and test 😄

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Add loading indicator when `ReconnectApp` is running
### What is the root cause of that problem?
This is a new feature
### What changes do you think we should make in order to solve the problem?
The test branch is here: https://github.com/nkdengineer/App/tree/fix/46611

1. We can create a `ProgressBar` component showing a loading indicator as the design above. Here is an example of this. We can refactor the logic and variables in the PR phrase.

```
import React, {useEffect, useState} from 'react';
import {View} from 'react-native';
import Animated, {Easing, useAnimatedStyle, useSharedValue, withTiming} from 'react-native-reanimated';

type ProgressBarProps = {
    shouldShow: boolean;
};

function ProgressBar({shouldShow}: ProgressBarProps) {
    const [isProgressing, setIsProgressing] = useState(false);
    const progressValue = useSharedValue(0);
    const [progress, setProgress] = useState(0);

    useEffect(() => {
        if (!shouldShow) {
            return;
        }

        setIsProgressing(true);
        setProgress(0);
    }, [shouldShow, setIsProgressing, setProgress]);

    useEffect(() => {
        let interval: NodeJS.Timeout;

        if (shouldShow) {
            setProgress(0);
            // eslint-disable-next-line react-compiler/react-compiler
            progressValue.value = 0;
            interval = setInterval(() => {
                setProgress((prev) => {
                    const nextProgress = prev + 10;
                    if (nextProgress >= 100) {
                        clearInterval(interval);
                        return 100;
                    }
                    return nextProgress;
                });
            }, 300);
        } else if (progress < 100) {
            progressValue.value = withTiming(100, {duration: 1000, easing: Easing.linear}, () => {
                setIsProgressing(false);
            });
            setProgress(100);
        }

        return () => {
            if (!interval) {
                return;
            }
            clearInterval(interval);
        };
    }, [shouldShow]);

    useEffect(() => {
        // eslint-disable-next-line react-compiler/react-compiler
        progressValue.value = withTiming(progress, {duration: 300});
    }, [progress, progressValue]);

    const animatedStyle = useAnimatedStyle(() => {
        return {
            width: `${progressValue.value}%`,
        };
    });

    if (!shouldShow && !isProgressing) {
        return null;
    }

    return (
        <View
            style={{
                height: 2,
                width: '100%',
                backgroundColor: '#1A3D32',
                borderRadius: 5,
                overflow: 'hidden',
            }}
        >
            <Animated.View style={[{height: '100%', backgroundColor: '#03D47C', borderRadius: 5}, animatedStyle]} />
        </View>
    );
}

ProgressBar.displayName = 'ProgressBar';

export default ProgressBar;

```

2. When `ReconnectApp` is called, we have `isLoadingReportData` in Onyx to know when it's running and it's complete. So we can use this variable to show the `ProgressBar` while `ReconnectApp` API is running.

- For `ReportScreen` get `isLoadingReportData` using `useOnyx` and show the `ProgressBar` below the header view if the screen is the large screen
```
const [isLoadingReportData] = useOnyx(ONYXKEYS.IS_LOADING_REPORT_DATA);
{!shouldUseNarrowLayout && <ProgressBar shouldShow={isLoadingReportData ?? false} />}
```
https://github.com/Expensify/App/blob/3666caa186f9c4530a1c57176eca601768c4953c/src/pages/home/ReportScreen.tsx#L755

- For LHN, we will show the `ProgressBar` below the `TopBar` component if the screen is small screen

```
const [isLoadingReportData] = useOnyx(ONYXKEYS.IS_LOADING_REPORT_DATA);
const {shouldUseNarrowLayout} = useResponsiveLayout();
{shouldUseNarrowLayout && <ProgressBar shouldShow={isLoadingReportData ?? false}/>}
```
https://github.com/Expensify/App/blob/3666caa186f9c4530a1c57176eca601768c4953c/src/pages/home/sidebar/SidebarScreen/BaseSidebarScreen.tsx#L56-L59


### What alternative solutions did you explore? (Optional)
NA

### Result



https://github.com/user-attachments/assets/9550a003-3ce5-404c-9571-3ab15e23843e


https://github.com/user-attachments/assets/ba29ded8-f3c1-4b43-8a78-2ef293145042

**Comment by dubielzyk-expensify** [comment]
Not overdue

**Comment by getusha** [comment]
The demos attached by both @ShridharGoel & @nkdengineer look way off to me...
Could you guys refer to the codepen provided by @dubielzyk-expensify https://codepen.io/dubielzyk/full/ExBVxVG?

**Comment by nkdengineer** [comment]
@getusha I updated my [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280) with the result as the expected design from codepen and also updated the test branch.

**Comment by nkdengineer** [comment]
I modified my [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280) with the expected.

**Comment by getusha** [comment]
Reviewing @nkdengineer's proposal

**Comment by JmillsExpensify** [comment]
How's the proposal review going?

**Comment by getusha** [comment]
I will provide an update today

**Comment by getusha** [comment]
I think it's better than the previous one, @nkdengineer. As you probably noticed, the timing isn't constant on which the width of the bar fills and goes away.  https://www.w3schools.com/cssref/tryit.php?filename=trycss_func_cubic-bezier

It's in the demo too:

>animation: 1.5s loading-bar .3s cubic-bezier(0.65, 0, 0.35, 1) infinite;


Could you check this and include in your proposal how we will implement the animation effect here?

**Comment by nkdengineer** [comment]
@getusha Updated [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280) with the detail about the animation.

The new result here

https://github.com/user-attachments/assets/19ba2f04-4861-42e0-80ff-6135e7fe4ec1

**Comment by dubielzyk-expensify** [comment]
This is looking great. Minor detail, but can we have it start from a place of empty, then fill? That is, it shouldn't start as fully green

**Comment by nkdengineer** [comment]
> This is looking great. Minor detail, but can we have it start from a place of empty, then fill? That is, it shouldn't start as fully green.

@dubielzyk-expensify Sure, it's expected from my solution. Here is the full result.

https://github.com/user-attachments/assets/0f34d0ba-ed49-4325-a11b-c21fc306ef1f

**Comment by flodnv** [comment]
> Coming from https://github.com/Expensify/App/issues/43453 we've landed on showing the thin bar loader at the top of the screen when ReconnectApp is running.

@dubielzyk-expensify sorry for realizing this so late -- why are we not showing the loader on the LHN in large screen mode?

**Comment by dannymcclain** [comment]
Dang this is feeling pretty darn slick!

>sorry for realizing this so late -- why are we not showing the loader on the LHN in large screen mode?

@shawnborton @flodnv @dubielzyk-expensify Good call. You're right that we had decided to initially go with the LHN on large screens. Relevant links:
- [Mocks](https://github.com/Expensify/App/issues/43453#issuecomment-2217098759)
- [Proposal to start with LHN implementation](https://github.com/Expensify/App/issues/43453#issuecomment-2217831358)
- Confirmation from [Shawn](https://github.com/Expensify/App/issues/43453#issuecomment-2217864940) & [Jon](https://github.com/Expensify/App/issues/43453#issuecomment-2219582012)

I'm honestly not sure how much I care re: LHN vs. Content pane, I think both look good. Just noting what we decided on that other issue 🙃

**Comment by shawnborton** [comment]
Yeah, I don't feel super strongly but I did initially like just using the LHN to start since it would be consistent everywhere. Does `ReconnectApp` really only modify what's in the LHN? Or does it also impact what is seen in the content pane?

**Comment by flodnv** [comment]
> Does ReconnectApp really only modify what's in the LHN? Or does it also impact what is seen in the content pane?

It could be both. If it modifies the content pane, it probably also modifies something in the LHN.

**Comment by getusha** [comment]
>I'm honestly not sure how much I care re: LHN vs. Content pane, I think both look good. Just noting what we decided on that other issue 🙃

So this means, on larger screens, display it only on the LHN. and on smaller screens, display it in both the LHN and the content pane. right?

**Comment by flodnv** [comment]
I think that makes the most sense to me.

**Comment by nkdengineer** [comment]
### Updated [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280) 

- Update the condition to show the loading indicator.

**Comment by shawnborton** [comment]
Hmm I think it could look weird if we have two different loading bars on desktop for the LHN and the content pane. I would rather we just pick one (and I would probably pick LHN?) or make the loading bar one continuous bar that stretches across the LHN and content pane at the same latitude.

**Comment by dubielzyk-expensify** [comment]
Yeah. Now that the proof of concept is out, I'd agree with those two solutions. I think across the whole thing would be the best though, but LHN works too.  I guess it's weird to see pulsing of skeleton and the bar in the same area

**Comment by dubielzyk-expensify** [comment]
Not overdue

**Comment by getusha** [comment]
@nkdengineer Instead of the bar suddenly disappearing, can we make it slide away? I think it would be better to render it all the time with a transparent background as well to avoid layout shift. WDYT @flodnv?

**Comment by nkdengineer** [comment]
> I think it would be better to render it all the time with a transparent background as well to avoid layout shift

@getusha If we display a transparent background after the API is complete, doesn't it suddenly disappearing?

We also want it to start from an empty place. So I think it's good to mount this component when the `ReconnectApp` API is in progress.
> This is looking great. Minor detail, but can we have it start from a place of empty, then fill? That is, it shouldn't start as fully green.

**Comment by getusha** [comment]
>@getusha If we display a transparent background after the API is complete, doesn't it suddenly disappearing?

I don't understand your question @nkdengineer, it will stay with a transparent background and we will just animate the green bar as needed.

**Comment by nkdengineer** [comment]
@getusha As my proposal we have a wrap view and the animated view (green line), do you mean we will always display the wrap view and only display the animated view when we need, right?

**Comment by getusha** [comment]
>@getusha As my proposal we have a wrap view and the animated view (green line), do you mean we will always display the wrap view and only display the animated view when we need, right?

@nkdengineer Yes, so when the `ReconnectApp` call ends, we will simply slide the green line away so it won't appear to be sudden.

**Comment by nkdengineer** [comment]
@getusha Got it, I updated my test branch and [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280). The new result:

https://github.com/user-attachments/assets/cf26c464-ce0c-4e3d-bf9c-3dc701ed70dc

**Comment by getusha** [comment]
>when the ReconnectApp call ends, we will simply slide the green line away so it won't appear to be sudden.

@nkdengineer can we also make sure the green bar slides away (set the width/left value with animation) instead of suddenly disappearing?

**Comment by nkdengineer** [comment]
@getusha Is this the expected behavior that we want?

https://github.com/user-attachments/assets/f0f364f8-1e55-43da-97dd-c641d7353943

**Comment by shawnborton** [comment]
Hmm it looks like there is a permanent 2px border now under the header area, which I don't think we want.

**Comment by getusha** [comment]
We could match the background? Otherwise it will appear jumpy when it gets
mounted and removed. Not sure but could also shift the layout for a moment,
which wouldn’t look ideal.

**Comment by shawnborton** [comment]
I think it should be absolutely positioned so it doesn't bump the content at all.

**Comment by dubielzyk-expensify** [comment]
> I think it should be absolutely positioned so it doesn't bump the content at all.

100% agree.

**Comment by getusha** [comment]
> I think it should be absolutely positioned so it doesn't bump the content at all.

That makes perfect sense!

**Comment by getusha** [comment]
@shawnborton @dubielzyk-expensify do you think the way we hide the bar matters? (i.e. adding a transition out/slide out instead of just disappearing suddenly?)

**Comment by shawnborton** [comment]
I think adding some kind of transition would be nice, whether it slides down from the top, or fades in, or something.

**Comment by dannymcclain** [comment]
Agree with you all about making it so the bar doesn't make the UI bump. Whether that's absolutely positioning it or using a transparent border as a placeholder, I don't have strong feelings. I just don't want a jumpy UI haha.

As for the way we hide it, I think a decently quick fade would be nice, because we don't really know at what point in its "bounce" it's going to go away, right?

**Comment by dubielzyk-expensify** [comment]
Agree with a pretty quick fade would look good here.

**Comment by dubielzyk-expensify** [comment]
Any updates?

**Comment by getusha** [comment]
@nkdengineer could you please update your proposal to align with https://github.com/Expensify/App/issues/46611#issuecomment-2312551255?

**Comment by nkdengineer** [comment]
Will update tomorrow.

**Comment by nkdengineer** [comment]
@shawnborton @dubielzyk-expensify   Is this the expected that we want?

https://github.com/user-attachments/assets/b56a6608-5883-433c-bf10-b96f815e1887

**Comment by shawnborton** [comment]
Did we decide to use two different loading bars here?
![CleanShot 2024-09-05 at 13 55 33@2x](https://github.com/user-attachments/assets/15b4b96a-b587-4bb1-9e49-a1903f92a4fd)

I thought we wanted to just use one on the LHN, but I could be mistaken. If we did use both, I think it's slightly odd that they are at different spots longitudinally - we could either move the LHN up slightly, or update the content header to always be 80px to match the LHN's header area.

**Comment by nkdengineer** [comment]
@shawnborton How about the animation of the loading indicator, is that good now?

**Comment by shawnborton** [comment]
I like how it fades out, but can we also make it fade in?

**Comment by nkdengineer** [comment]
> but can we also make it fade in?

@shawnborton Yes, we can also make fade-in like this video. The timing fade-in/out is `300ms`.

https://github.com/user-attachments/assets/cdebb178-10e3-4193-8291-a9e93049c0a8

**Comment by shawnborton** [comment]
Nice, something like that feels better to me. Let's get clarification on if we're doing just one or two of these on desktop as well. cc @Expensify/design for more thoughts and feedback!

**Comment by dannymcclain** [comment]
Love the fading. 🙌

>I thought we wanted to just use one on the LHN, but I could be mistaken.

[This is what we decided](https://github.com/Expensify/App/issues/46611#issuecomment-2293578862): only use one on the LHN. The chat/main content pane would only get it on mobile (where that's the only thing you can see).

>on larger screens, display it only on the LHN. and on smaller screens, display it in both the LHN and the content pane.

https://github.com/Expensify/App/issues/46611#issuecomment-2296020605 ☝️

**Comment by nkdengineer** [comment]
Thanks for the confirmation.

@getusha Updated [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280).

**Comment by flodnv** [comment]
Love it! 👍

**Comment by getusha** [comment]
Looks good to me as well,
We can proceed with @nkdengineer's [proposal](https://github.com/Expensify/App/issues/46611#issuecomment-2262387280)
🎀 👀 🎀 C+ Reviewed

**Comment by jliexpensify** [comment]
Payment Summary

- C: @nkdengineer $250
- C+: @getusha $250 (ND payment)

Upworks job - https://www.upwork.com/jobs/~021836556152855644897

**Comment by blimpich** [comment]
Waiting on @getusha to review

**Comment by jliexpensify** [comment]
Just a heads up I'm OOO from 3rd to 14th, but I don't think anything is needed from me during this period as we're waiting on @getusha. Feel free to reassign to another B0 for Upworks payment if needed though, otherwise I'll review and pay when I get back!

**Comment by blimpich** [comment]
Reassigning to @srikarparsi since I'll be OOO

**Comment by jliexpensify** [comment]
Not overdue, PR being reviewed.

**Comment by trjExpensify** [comment]
Can payment be processed here?

**Comment by nkdengineer** [comment]
@trjExpensify The [PR](https://github.com/Expensify/App/pull/48772) is still reviewing.

**Comment by jliexpensify** [comment]
Removed you @shubham1206agra

**Comment by jliexpensify** [comment]
Looks like @srikarparsi has to address a comment in the PR

**Comment by jliexpensify** [comment]
I think we're now waiting on @getusha

**Comment by srikarparsi** [comment]
PR in review

**Comment by iwiznia** [comment]
Seems this PR caused various regressions, see https://github.com/Expensify/App/pull/52112

**Comment by nkdengineer** [comment]
Noted to check in the next PR.

**Comment by srikarparsi** [comment]
To leave a quick update, the [initial PR](https://github.com/Expensify/App/pull/48772#issuecomment-2409001095) caused multiple regressions so a new [PR](https://github.com/Expensify/App/pull/52272) is being worked on.

**Comment by srikarparsi** [comment]
@nkdengineer do you have an ETA on when the new PR will be ready?

**Comment by nkdengineer** [comment]
I'm investigating to avoid the previous regression, will give an update today or tomorrow.

**Comment by nkdengineer** [comment]
@getusha @srikarparsi we have a open follow PR [here](https://github.com/Expensify/App/pull/52272)

**Comment by muttmuure** [comment]
How close are we with this one? Can we get it out before EoW?

**Comment by jliexpensify** [comment]
@srikarparsi @getusha @muttmuure - just a heads up that I am OOO from 19th - 29th December. If this needs payment, please feel free to reassign, otherwise I'll review and sort out when I'm back.

**Comment by srikarparsi** [comment]
> How close are we with this one? Can we get it out before EoW?

Hey @muttmuure, we're waiting on [deciding](https://github.com/Expensify/App/pull/52272#issuecomment-2552698993) if we want the background of the loading indicator to be 1px or 2px tall and we'll make the necessary changes. I think towards the beginning of next week is probably a good estimate to get this merged.

**Comment by adhorodyski** [comment]
Hey everyone, I just stumbled upon this thread because I wanted to propose the ~same thing for a more native UX. 

One question, how does this loading indicator play together with the skeletons on LHN? Ideally, we could get rid of them with this feature introduced.

**Comment by srikarparsi** [comment]
Yeah, slack would be perfect. This was the [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1699972596342559) this was originally discussed so that should be a good place :)

**Comment by flodnv** [comment]
That's great news! But why do I not see it on Desktop? I'm on v9.0.91-0

**Comment by jliexpensify** [comment]
Payment Summary

- C: @nkdengineer $250 (to be paid via Upworks) - please accept the new offer
- C+: @getusha $250 (to be paid vis ND requests) - is there a checklist to be completed?

Upworks job: https://www.upwork.com/jobs/~021886616910383815201

**Comment by jliexpensify** [comment]
Bumping @getusha and @nkdengineer for this [comment](https://github.com/Expensify/App/issues/46611#issuecomment-2632735050)

**Comment by jliexpensify** [comment]
Paid and job closed!

@getusha:

![Image](https://github.com/user-attachments/assets/76f95d5a-de91-4c6b-a16f-0c1d76bf29bc)

**Comment by jliexpensify** [comment]
Bumping @getusha for this:

![Image](https://github.com/user-attachments/assets/4c09b157-2cd3-4bd0-8fdf-f5ee05e78f96)

**Comment by jliexpensify** [comment]
@getusha bumping for the checklist again

![Image](https://github.com/user-attachments/assets/bd8b4a01-adb1-422c-b229-88a0cc068832)

**Comment by jliexpensify** [comment]
Bumping @getusha once again for a checklist

![Image](https://github.com/user-attachments/assets/4169062a-f4b6-40f9-9ea3-4501b61c5440)

**Comment by jliexpensify** [comment]
@getusha checklist plzzzzzzzzzzzzz:

![Image](https://github.com/user-attachments/assets/66762635-fed6-46f4-b260-e15261912c7c)

**Comment by jliexpensify** [comment]
@getusha checklist plzzzzz:

![Image](https://github.com/user-attachments/assets/005d0254-f561-48e1-8d9b-bb59a036b071)

**Comment by jliexpensify** [comment]
![Image](https://github.com/user-attachments/assets/0bf27d45-c924-434d-8d29-d148342b288d)

**Comment by jliexpensify** [comment]
🤷 a meme a day until I see a checklist

**Comment by s77rt** [comment]
@getusha Do not do the checklist, those memes are funny

**Comment by getusha** [comment]
Oh my gosh!
I mostly try to complete multiple checklists at once when submitting ND request. Can't believe i missed all this😄 sorry!

**Comment by getusha** [comment]
ANDDD this is a new feature we don't need a checklist. here is a regression test proposal.

## Regression Test Proposal

### Test:


1. Open the app
2. Force offline
3. Navigate to Settings > Troubleshoot
4. Clear cache and restart
5. Navigate to Inbox
6. Verify that there is a loading indicator bar on top of the LHN

Do we agree 👍 or 👎

**Comment by getusha** [comment]
Me when it's time to fill out the checklist:


<img src="https://github.com/user-attachments/assets/304f8a33-6be0-4138-9e71-f859a69bd9f8" width="300"/>

**Comment by jliexpensify** [comment]
@getusha I was wondering if you were OOO or just ignoring me 😅 

![Image](https://github.com/user-attachments/assets/716272d5-b34b-49be-9d5b-8a179dbe39cf)

^ Final meme, thanks everyone for participating

**Comment by garrettmknight** [comment]
$250 approved for @getusha after almost an entire year.
